samza-core/src/main/java/org/apache/samza/coordinator/RunIdGenerator.java

```diff
@@ -0,0 +1,114 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Generates RunId for Standalone use case
+ * If there is only one processor in the quorum (registered with ClusterMembership) then create new runid and add to store
+ * Else read runid from the store
+ *
+ * Steps to generate:
+ * 1. acquire lock
+ * 2. add self to quorum (register itself with ClusterMembership)
+ * 3. get number of processors in quorum
+ * 4. if qurorum size is 1 (only self) then create new runid and write to store
+ * 5. if quorum size if greater than 1 then read runid from store
+ * 6. unlock
+ */
+public class RunIdGenerator {
+  private static final Logger LOG = LoggerFactory.getLogger(RunIdGenerator.class);
+
+  private final CoordinationUtils coordinationUtils;
+  private final MetadataStore metadataStore;
+  private ClusterMembership clusterMembership = null;
+
+  public RunIdGenerator(CoordinationUtils coordinationUtils, MetadataStore metadataStore) {
+    Preconditions.checkNotNull(coordinationUtils, "CoordinationUtils cannot be null");
+    Preconditions.checkNotNull(metadataStore, "MetadataStore cannot be null");
+    this.coordinationUtils = coordinationUtils;
+    this.metadataStore = metadataStore;
+  }
+
+  public String getRunId() {
+    DistributedLock runIdLock;
+    String runId = null;
+
+
+    try {
+      runIdLock = coordinationUtils.getLock(CoordinationConstants.RUNID_LOCK_ID);
+      clusterMembership = coordinationUtils.getClusterMembership();
+      if (runIdLock == null || metadataStore == null || clusterMembership == null) {
+        throw new SamzaException("Failed to create utils for run id generation");
+      }
+    } catch (Exception e) {
+      LOG.warn(e.getMessage());
+      return null;
+    }
+
+    try {
+      // acquire lock to write or read run.id
+      if (runIdLock.lock(CoordinationConstants.LOCK_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+        LOG.info("lock acquired for run.id generation by this processor");
+        clusterMembership.registerProcessor();
+        int numberOfActiveProcessors = clusterMembership.getNumberOfProcessors();
+        if (numberOfActiveProcessors == 0) {
+          String msg = String.format("Processor failed to fetch number of processors for run.id generation");
+          throw new SamzaException(msg);
+        }
+        if (numberOfActiveProcessors == 1) {
+          runId =
+              String.valueOf(System.currentTimeMillis()) + "-" + UUID.randomUUID().toString().substring(0, 8);
+          LOG.info("Writing the run id for this run as {}", runId);
+          metadataStore.put(CoordinationConstants.RUNID_STORE_KEY, runId.getBytes("UTF-8"));
+        } else {
+          runId = new String(metadataStore.get(CoordinationConstants.RUNID_STORE_KEY));
+          LOG.info("Read the run id for this run as {}", runId);
+        }
+      }
+    } catch (TimeoutException e) {
+      String msg = String.format("Processor timed out waiting to acquire lock for run.id generation");
```

Comment:
Seems like a plain string. Why use `String.format`?
Can we eliminate the interim local variable and have it plugged in directly to the exception creation?